Support managed debugging when CET is enabled

tommcdon opened the pull request:
This change adds out of process SetThreadContext support on Windows x64.  The feature will be automatically enabled if the process is enabled for CET or with a configuration setting (useful for testing purposes).  I've validated our mainline scenarios with the Visual Studio debugger, including breakpoints, stepping, mixed-mode debugging, and Edit-And-Continue (Hot Reload).

tommcdon committed:
Support managed debugging when CET is enabled

mangod9 commented on src/coreclr/debug/ee/debugger.cpp:
```diff
@@ -16631,4 +16653,95 @@ void Debugger::StartCanaryThread()
+    success = InitializeContext(pBuffer, contextFlags, &pContext, &contextSize);
+    if (!success)
+    {
+        _ASSERTE(!"InitializeContext failed");
```
does pBuffer need to be deleted for the couple of error paths?

janvorli replied:
No, it is allocated by alloca on stack.

janvorli commented on src/coreclr/vm/encee.cpp:
```diff
@@ -835,7 +835,14 @@ NOINLINE void EditAndContinueModule::FixContextAndResume(
-    ClrRestoreNonvolatileContextWorker(pContext, ssp);
+    if (g_pDebugInterface->IsOutOfProcessSetContextEnabled())
+    {
+        g_pDebugInterface->SendSetThreadContextNeeded(pContext);
```
Have you hit a problem where the ClrRestoreNonvolatileContextWorker was not working here? I have thought it would be sufficient as it doesn't use any Windows APIs for restoring the context and thus we should be able to restore the context to any target Rip.

tommcdon replied:
Yes this AV's in the EnC code path.

janvorli replied:
That's strange then. Would you mind getting me details on the AV (stack trace, ...)? I would like to understand what exactly was wrong.
With your change, could you also please remove the ssp locals from this function, as it is not needed anymore (it is needed only for CET enabled and in that case, we use this new added code)? You'd also need to change the `ClrRestoreNonvolatileContextWorker(pContext, ssp);` to `ClrRestoreNonvolatileContext(pContext);`

tommcdon replied:
> Would you mind getting me details on the AV (stack trace, ...)? I

The stack is empty and RIP is 0.  Further investigation will be needed to understand why ClrRestoreNonvolatileContext does not work in this scenario.  I'll move forward with the existing change for .NET 7 while we continue to investigate this issue.

mikem8361 commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -7535,6 +7535,31 @@ HRESULT DacDbiInterfaceImpl::EnableGCNotificationEvents(BOOL fEnable)
+
+        if (FAILED(hr))
+        {
+            ThrowHR(CORDBG_E_READVIRTUAL_FAILURE);
```
Why are you throwing an exception here and then catching below? When you could just `return CORDBG_E_READVIRTUAL_FAILURE`?

mikem8361 commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -7535,6 +7535,31 @@ HRESULT DacDbiInterfaceImpl::EnableGCNotificationEvents(BOOL fEnable)
+
+        if (cbRead != size)
+        {
+            ThrowWin32(ERROR_PARTIAL_COPY);
```
Can't this just `return HRESULT_FROM_WIN32(ERROR_PARTIAL_COPY);` ?

mikem8361 commented on src/coreclr/debug/daccess/dacdbiimpl.cpp:
```diff
@@ -7535,6 +7535,31 @@ HRESULT DacDbiInterfaceImpl::EnableGCNotificationEvents(BOOL fEnable)
+            ThrowWin32(ERROR_PARTIAL_COPY);
+        }
+    }
+    EX_CATCH_HRESULT(hr);
```
If you change the above throws to returns, then you don't need the EX_TRY/EX_CATCH_HRESULT.  ReadVirtual should never throw an exception.

mikem8361 commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -11377,7 +11414,15 @@ HRESULT CordbProcess::Filter(
             // holder will invoke DeleteIPCEventHelper(pManagedEvent).
         }
+#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+        else if (dwFirstChance && pRecord->ExceptionCode == STATUS_BREAKPOINT && pRecord->ExceptionAddress == m_runtimeOffsets.m_setThreadContextNeededAddr) /*|| pRecord->ExceptionCode == STATUS_SINGLE_STEP*/
```
Can you remove the commented out code: `/*|| pRecord->ExceptionCode == STATUS_SINGLE_STEP*/`

noahfalk commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -7719,6 +7721,7 @@ HRESULT CordbProcess::GetRuntimeOffsets()
             m_runtimeOffsets.m_signalHijackCompleteBPAddr,
             m_runtimeOffsets.m_excepNotForRuntimeBPAddr,
             m_runtimeOffsets.m_notifyRSOfSyncCompleteBPAddr,
+            m_runtimeOffsets.m_setThreadContextNeededAddr,
```
Lets move the #ifdef FEATURE_INTEROP_DEBUGGING so it only covers the first six flares and get rid of the comment that says flares are only for interop debugging.

noahfalk commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -7659,6 +7659,8 @@ HRESULT CordbProcess::GetRuntimeOffsets()
          m_runtimeOffsets.m_debuggerWordTLSIndex));
+    LOG((LF_CORDB, LL_INFO10000, "    m_setThreadContextNeededAddr=     0x%p\n",
+         m_runtimeOffsets.m_setThreadContextNeededAddr));
 #endif // FEATURE_INTEROP_DEBUGGING
```
Lets move the log statement outside the ifdef FEATURE_INTEROP_DEBUGGING

noahfalk reviewed (approved):
LGTM modulo comments inline

tommcdon commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -11152,7 +11155,41 @@ void CordbProcess::FilterClrNotification(
     }
 }
 
+#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
```
> I'd recommend using an ifdef named with the feature rather than the OS or architecture. Otherwise when we need to port features in the future it is harder to determine which code matters. You would need to add the new define if one doesn't already exist.

Thanks for the suggestion.  @hoyosjs suggested the same.  It looks like CET is built into the product now and does not have a define.  I believe it is beyond the scope of this PR to add a new CET ifdef.  I will create an issue to track this request.

noahfalk replied:
I don't think there is much ceremony here, you could add the define with less text than it takes to file the tracking issue : )

In src\coreclr\clrdefinitions.cmake add:
```
if (CLR_CMAKE_TARGET_WIN32 AND CLR_CMAKE_TARGET_ARCH_AMD64)
    add_definitions(-DFEATURE_CET)
endif (CLR_CMAKE_TARGET_WIN32 AND CLR_CMAKE_TARGET_ARCH_AMD64)
```

tommcdon replied:
Yes I had already experimented with adding a feature flag for out-of-process setthreadcontext - https://github.com/tommcdon/runtime/commit/b8ad42d4182d3e7a8212bcd6a6d28180b4b082e1.  I had originally felt the new define wasn't necessary because 1) we need architecture specific ifdef's under the feature ifdef's even though the feature flag is only enabled for one platform/one architecture, 2) the runtime's CET implementation doesn't have a feature flag so seems confusing to a add a feature flag that only works on part of the runtime's codebase, 3) the feature is a runtime check (not compile time) for x64 Windows.  With that said I don't feel strongly enough to push back, so I'll move forward with adding that change :).

noahfalk replied:
The suggestion is largely based on the assumption that even if CET is windows x64 for now, its probably only a matter of time before we need to support it on another OS or another architecture. When that time comes if devs can search for FEATURE_CET in the source or enable the ifdef and see what breaks that is much easier than having to do lots of code review figuring out which #ifdef WINDOWS X64 code is CET and which is not : )

tommcdon commented on src/coreclr/debug/di/process.cpp:
```diff
@@ -11152,7 +11155,41 @@ void CordbProcess::FilterClrNotification(
+        ThrowHR(hr);
+    }
+
+    SetLiveContext(dwThreadId, pContext);
```
> Instead of a new function this can GetDataTarget()->SetThreadContext(...)

Good suggestion.  Our implementation of ICorDebugMutableDataTarget only handles the legacy context scenario.  We either need to change it or add our own SetThreadContext.  I opted to create our own as I wanted minimize risk of the change and not alter existing debugging code (since we are close to the end of .NET 7).  I'll create a tracking item for .NET 8 to unify these code paths.

noahfalk replied:
In that case I'd suggest adding a comment in the code describing what is different between the new function and the data target variant and that we'd like to fold them back together again. I expect the most likely result of a 'refactor this later' work item is that it will be punted to Future and closed without ever doing it, but at least with the comment future folks will know whats up and they might do the refactoring if it is convenient.

tommcdon committed:
Use DataTarget ReadVirtual function to read data
Remove RS GetLiveContext/SetLiveContext implementations
Update debugger contract

tommcdon commented on src/coreclr/debug/ee/debugger.cpp:
```diff
@@ -16631,4 +16653,95 @@ void Debugger::StartCanaryThread()
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
```
The GC_NOTRIGGER macro calls [IncrementGCForbidCount](https://github.com/dotnet/runtime/blob/2201016c1e13bdb9abf49e2e38cadf4ee0568df2/src/coreclr/inc/contract.h#L649) which doesn't document because we SetThreadContext while inside of the method.  I don't know if it is possible to scope the contract to only part of the method.

noahfalk replied:
you can probably use STATIC_CONTRACT_GC_NOTRIGGER

tommcdon committed:
Clean up contracts and add extra validation in SetThreadContextNeeded message

tommcdon committed:
Add OUT_OF_PROCESS_SETTHREADCONTEXT ifdef

tommcdon committed:
Add todo comment to merge ICorDebugMutableDataTarget::SetThreadContext  with HandleSetThreadContextNeeded implementation

tommcdon commented:
Test failures look unrelated to this PR and so merging
